plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/p5edges/orthogonal/HyperEdgeSegmentSplitter.java

```diff
@@ -0,0 +1,234 @@
+        Stream<Double> outCoordinates = segments.stream().flatMap(s -> s.getOutgoingConnectionCoordinates().stream());
+        
+        double[] sortedCoordinates = Streams.concat(inCoordinates, outCoordinates)
+            .mapToDouble(coordinate -> coordinate.doubleValue())
+            .sorted()
+            .toArray();
+        
+        // Go through each pair of coordinates and create free areas for those that are at least twice the critical
+        // threshold
+        for (int i = 1; i < sortedCoordinates.length; i++) {
+            if (sortedCoordinates[i] - sortedCoordinates[i - 1] >= 2 * criticalConflictThreshold) {
+                freeAreas.add(new FreeArea(
+                        sortedCoordinates[i - 1] + criticalConflictThreshold,
+                        sortedCoordinates[i] - criticalConflictThreshold));
+            }
+        }
+        
+        return freeAreas;
+    }
+    
+    ////////////////////////////////////////////////////////////////////////////////////////////////////////////
+    // Split Segment Decisions
+
+    /**
+     * Given a list of hyperedge dependencies, this method assembles a list of {@link HyperEdgeSegment}s which, if
+     * split, will cause the critical cycles caused by the dependencies to be broken.
+     */
+    private LinkedHashSet<HyperEdgeSegment> decideWhichSegmentsToSplit(
+            final List<HyperEdgeSegmentDependency> dependencies) {
+        
+        // We want to be sure to have a stable iteration order, but we also want fast containment checks
+        LinkedHashSet<HyperEdgeSegment> segmentsToSplit = new LinkedHashSet<>();
+        
+        for (HyperEdgeSegmentDependency dependency : dependencies) {
+            HyperEdgeSegment sourceSegment = dependency.getSource();
+            HyperEdgeSegment targetSegment = dependency.getTarget();
+            
+            // If either of the involved segments were already selected for splitting because of another dependency,
+            // that's sufficient
+            if (segmentsToSplit.contains(sourceSegment) || segmentsToSplit.contains(targetSegment)) {
+                continue;
+            }
+            
+            // One segment will be split, and the other one will be remembered to have caused the split. The latter
+            // will have to stay between the two segments the former will be split into, or else we'll have overlaps
+            // again
+            HyperEdgeSegment segmentToSplit = sourceSegment;
+            HyperEdgeSegment segmentCausingSplit = targetSegment;
+            
+            // This is perfectly fine unless the segment to split represents a hyperedge and the other does not. We
+            // prefer splitting regular edges since hyperedges have a higher chance of causing additional crossings.
+            // Note: the original code took the layout direction into account here. We may want to investigate at some
+            // point whether that's necessary.
```

Comment:
Did we investigate that?

Reply:
Nope. This is more like, "If it turns out to look odd, which I haven't seen so far, this would be a place to look."